### ISSUE TITLE
regresion: parsing failure in derive expansion?

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#115775]
```
[INFO] [stdout] error: expected one of: `for`, parentheses, `fn`, `unsafe`, `extern`, identifier, `::`, `<`, square brackets, `*`, `&`, `!`, `impl`, `_`, lifetime
[INFO] [stdout]  --> tests/happy.rs:9:13
[INFO] [stdout]   |
[INFO] [stdout] 9 |     thing: &'static str,
[INFO] [stdout]   |             ^^^^^^^
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: expected one of `>`, a const expression, lifetime, or type, found `'`
[INFO] [stdout]  --> examples/simple.rs:3:17
[INFO] [stdout]   |
[INFO] [stdout] 3 | #[derive(Debug, Display)]
[INFO] [stdout]   |                 ^^^^^^^
[INFO] [stdout]   |                 |
[INFO] [stdout]   |                 expected one of `>`, a const expression, lifetime, or type
[INFO] [stdout]   |                 while parsing this item list starting here
[INFO] [stdout]   |                 the item list ends here
[INFO] [stdout]   |
[INFO] [stdout]   = note: this error originates in the derive macro `Display` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: aborting due to previous error
```

https://crater-reports.s3.amazonaws.com/beta-1.73-1.2/beta-2023-09-10/reg/displaydoc-watt-0.1.0/log.txt

**Comment by Mark-Simulacrum** [comment]
Possibly related (if not, let's file a new issue for it):

* https://crater-reports.s3.amazonaws.com/beta-1.73-1.2/beta-2023-09-10/reg/momo-0.2.2/log.txt

**Comment by lukas-code** [comment]
Regression in https://github.com/rust-lang/rust/pull/114617, where the order of the variants of [`proc_macro::Spacing`](https://doc.rust-lang.org/nightly/proc_macro/enum.Spacing.html) got changed.

MCVE: [(playground)](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=78508e47bee43d0df3f12ddd68d27401)
```rust
extern crate proc_macro;

fn main() {
    assert_eq!(proc_macro::Spacing::Alone as u8, 0);
}
```

The `watt` crate is relying on the discriminant values here:

https://github.com/dtolnay/watt/blob/8dc8981c6c5c2a3e0ff6c9ec3830133538c78870/src/encode.rs#L108-L112

https://github.com/dtolnay/watt/blob/8dc8981c6c5c2a3e0ff6c9ec3830133538c78870/src/decode.rs#L117-L124

@rustbot label -E-needs-bisection -E-needs-mcve +S-has-mcve +A-proc-macros

**Comment by apiraino** [comment]
Just to understand the context: will this be considered a breaking change (in some form)? If yes, is something users should be aware of in the lreease notes?

**Comment by Mark-Simulacrum** [comment]
I think flipping the order wasn't necessary for rustc, so we should probably undo that and backport.

Cc @dtolnay as well since you might want to stop depending on this unstable (imo) detail

**Comment by dtolnay** [comment]
Yeah this is a watt bug. I don't think the rustc change needs to be reverted.

**Comment by dtolnay** [comment]
I published fixes in watt 0.3.1 and 0.4.5.

**Comment by Mark-Simulacrum** [comment]
Closing since the watt fixes are out.
